File: 611div3/c.cpp
#include <iostream>
#include <unordered_set>
#include <vector>

using namespace std;

int N;
int friends[200010];

int main() {
    cin >> N;
    for (int i = 0; i < N; i++)
        cin >> friends[i];
    unordered_set<int> connect;
    unordered_set<int> available;
    for (int i = 1; i <= N; i++) { available.insert(i); }
    for (int i = 0; i < N; i++) {
        if (friends[i] == 0)
            connect.insert(i + 1);
        else
            available.erase(friends[i]);
    }
    vector<int> same, diff, wait;
    for (const auto& it : connect) {
        if (available.count(it))
            same.push_back(it);
        else
            diff.push_back(it);
    }
    
    for (const auto& it : available) {
        if (!connect.count(it))
            wait.push_back(it);
    }


    for (int i = 0; i < same.size(); i++) {
        friends[same[i] - 1] = i == 0 ? same.back() : same[i - 1];
    }
    for (int i = 0; i < diff.size(); i++) {
        friends[diff[i] - 1] = wait[i];
    }
    if (same.size() == 1) {
        swap(friends[same[0] - 1], friends[diff[0] - 1]);
    }
    for (int i = 0; i < N; i++)
        cout << friends[i] << " ";
    return 0;
}

File: 611div3/c.py
n = int(input())

friend = list(map(int, input().split()))

available = {}
connect = {}

for i in range(1, n + 1):
    if friend[i - 1] == 0:
        connect[i] = 1
    if i not in friend:
        available[i] = 1

same = []
diff = []

for i in connect:
    if i in available:
        same.append(i)
        del available[i]
    else:
        diff.append(i)

print(same)
print(diff)

if (len(same) != 1):
    for i in range(len(same)):
        friend[same[i]] = same[i - 1]
    for i in available:
        friend[diff[-1]] = i
        del diff[-1]
else:
    available[same[i]] = 1
    cnt = 0
    for i in available:
        frient[diff[cnt]] = i
        cnt += 1
        

ans = ""

for i in friend:
    ans += str(i) + " "

print(ans)